### PR TITLE
convert the implementation of the action to Dart

### PR DESCRIPTION
**Pull request by devoncarew**
Convert the implementation of the action to Dart:

- a mostly direct translation from the Bash implementation to Dart
- kept the `dart.yml` CI the same, so we could verify correct behavior of the port
- as part of the port, addressed an issue related to auto-detection of the platform architecture (fix https://github.com/dart-lang/setup-dart/issues/59)
- also, added a new action output - `dart-version`
- done as part of https://github.com/dart-lang/setup-dart/issues/62
- updated the changelog w/ the relevant changes, and rev'd the action version to `1.5.0-dev`
- updated with our std CONTRIBUTING.md file from dart-lang/.github
- added a DEVELOPING.md file, with some info about how the action is built and how to work on it

### CODE REVIEW

File: example/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+.dart_tool/
```
Comment (kevmoo) [inline example/.gitignore:1]: just delete this, right?

Comment (devoncarew) [reply example/.gitignore:2]: I think it's ok to keep this around (example's ~self-sufficient).

**Comment by devoncarew** [comment]
@joshualitt + @srujzs - can you review the js interop code (lib/node/*), the bootstrap script (lib/main.mjs), and anything else that looks interesting? Thanks!

**Review by joshualitt — APPROVED** [review]
The JS interop code looks great. I don't think there is anything that needs to change in this CL. One thing for us to ponder as team, there is some overlap in the JS interop used in this package and `package:web`. It may make sense for us to publish a node interop package so we can share this work. Perhaps some of our users may also want to use such a package.

Anyway, food for thought. Regardless, that is far outside the scope of this CL.

**Comment by devoncarew** [comment]
Thanks! cc'ing @mit for fyi on the PR; @athomas - do you mind doing a general review of this PR? It's substantially the same as the previous one (and trying to keep minimal diffs from the current bash impl).

File: lib/node/actions/http_client.dart [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:js_interop';
+
+import 'package:js/js.dart' as js;
+
+@JS()
+@js.staticInterop
+class HttpClient {
+  external factory HttpClient(
```
Comment (srujzs) [inline lib/node/actions/http_client.dart:12]: Up to you if you want to change the parameter types here to JS types.

**Review by srujzs — APPROVED** [review]
LGTM too!

File: lib/node/actions/http_client.dart (hunk 2 above)
Comment (devoncarew) [reply lib/node/actions/http_client.dart:12]: sgtm - updates to use JSString and JSArray

File: lib/main.dart [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    // architecture
+    var architecture = core.getInput('architecture');
+    if (architecture.isEmpty) {
+      architecture = getArch();
+    }
+
+    // calculate version and channel
+    String version;
+    String channel;
+
```
Comment (athomas) [inline lib/main.dart:48]: super nit: this newline seems inconsistent with the style above

File: lib/main.dart [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        'the $channel ($flavor) channel.');
+
+    // Calculate url based on https://dart.dev/tools/sdk/archive#download-urls.
+    final url = 'https://storage.googleapis.com/dart-archive/'
+        'channels/$channel/$flavor/$version/sdk/'
+        'dartsdk-$os-$architecture-release.zip';
+
+    // Use a cached sdk or download and cache the sdk; using a 'raw' sdk flavor
+    // disables caching.
+    final toolName = flavor == 'raw' ? 'dart_raw' : 'dart';
```
Comment (athomas) [inline lib/main.dart:83]: Replace `flavor == 'raw'` with `raw` like in the next line.

File: lib/main.dart [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+String? getVersionFromSdk(String sdkPath) {
+  final versionFilePath = path.join(sdkPath, 'version');
+  print(versionFilePath);
+
+  if (fs.existsSync(versionFilePath)) {
+    return fs.readFileSync(versionFilePath, 'utf8').trim();
+  } else {
+    return null;
```
Comment (athomas) [inline lib/main.dart:132]: Given that the one place that calls this asserts not-null, maybe this should be non-nullable instead and unconditionally read the file? We still might get random IO related errors from fs.readFileSync and the file doesn't exist we get a runtime error anyway.

File: lib/main.dart [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+  final http = HttpClient('setup-dart', [], {
+    'allowRedirects': true,
+    'maxRedirects': 3,
+    'allowRetries': true,
+    'maxRetries': 3,
+  });
+
+  JSObject response = await promiseToFuture(http.getJson(url));
+  JSObject? result = getProperty(response, 'result');
+  return result == null ? null : getProperty(result, 'version');
```
Comment (athomas) [inline lib/main.dart:188]: Given the call sites assert non-nullability, we could just make this non-nullable and move the assert here.

File: .github/dependabot.yaml [hunk 7]
```diff
@@ -2,7 +2,12 @@
 version: 2
```
Comment (athomas) [inline .github/dependabot.yaml:3]: Just curious, why no package-ecosystem: pub?

**Review by athomas — APPROVED** [review]
LGTM, didn't review JS interop because people that know what they're doing already did :)

File: .github/dependabot.yaml (hunk 7 above)
Comment (devoncarew) [reply .github/dependabot.yaml:3]: I ran into some issues when eval'ing it when it was in beta (https://github.com/dependabot/dependabot-core/issues/4979), and was generally waiting for them to be addressed before using the functionality more widely.

File: lib/main.dart (hunk 4 above)
Comment (devoncarew) [reply lib/main.dart]: done

File: lib/main.dart (hunk 5 above)
Comment (devoncarew) [reply lib/main.dart]: sounds good; makes sense to not expect a missing file

File: lib/main.dart (hunk 6 above)
Comment (devoncarew) [reply lib/main.dart]: +1 - I changed this to return a non-null value, and instead throw an exception if we can't retrieve the latest version

File: lib/main.dart (hunk 3 above)
Comment (devoncarew) [reply lib/main.dart:48]: Hmm, I don't see anything different about it? Feel free to use the 'add a suggestion' feature to edit the file here.

**Review by devoncarew — COMMENTED** [review]
Thanks! PR updated.

File: lib/main.dart (hunk 3 above)
Comment (athomas) [reply lib/main.dart:48]: I meant to remove the newline, sorry for being unclear:

All the other blocks where:
<declaractions>
<if statement>

Except this which is:

<declarations>
<newline>
<if statement>

Not important though, so let's just resolve this.

**Pull request merged**